src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+ *
+ * It plans, manages and executes the cleanup jobs asynchronously in a separate thread so cleaning up does not affect
+ * the performance of the other tasks of the node.
+ */
+public class GarbageCollector {
+    /**
+     * The interval in milliseconds that the garbage collector will check if new cleanup tasks are available.
+     */
+    private static final int GARBAGE_COLLECTOR_RESCAN_INTERVAL = 10000;
+
+    /**
+     * Logger for this class allowing us to dump debug and status messages.
+     */
+    private static final Logger log = LoggerFactory.getLogger(GarbageCollector.class);
+
+    /**
+     * Boolean flag that indicates if the node is being shutdown.
+     */
+    boolean shuttingDown = false;
+
+    /**
+     * Holds a reference to the tangle instance which acts as an interface to the used database.
+     */
+    final Tangle tangle;
+
+    /**
+     * Holds a reference to the {@link SnapshotManager} that this garbage collector belongs to.
+     */
+    final SnapshotManager snapshotManager;
+
+    /**
+     * Holds a reference to the {@link TipsViewModel} which is necessary for removing tips that were pruned.
+     */
+    final TipsViewModel tipsViewModel;
+
+    /**
+     * A map of {@link JobParser}s allowing us to determine how to parse the jobs from the garbage collector state file,
+     * based on their type.
+     */
+    private final HashMap<String, JobParser> jobParsers = new HashMap<>();
+
+    /**
+     * A map of {@link QueueProcessor}s allowing us to process queues based on the type of the job.
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueProcessor> queueProcessors = new HashMap<>();
+
+    /**
+     * A map of {@link QueueConsolidator}s allowing us to consolidate the queues to consume less "space".
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueConsolidator> queueConsolidators = new HashMap<>();
+
+    /**
+     * List of cleanup jobs that shall get processed by the garbage collector (grouped by their class).
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> garbageCollectorJobs = new HashMap<>();
+
+    /**
+     * The constructor of this class stores the passed in parameters for future use and restores the previous state of
+     * the garbage collector if there is a valid one (to continue with cleaning up after IRI restarts).
+     *
+     * Before restoring the garbage collector state it registers the available job types, so we know how to parse and
+     * process the found jobs.
+     */
+    public GarbageCollector(Tangle tangle, SnapshotManager snapshotManager, TipsViewModel tipsViewModel) {
+        this.tangle = tangle;
+        this.snapshotManager = snapshotManager;
+        this.tipsViewModel = tipsViewModel;
+
+        MilestonePrunerJob.registerInGarbageCollector(this);
+        UnconfirmedSubtanglePrunerJob.registerInGarbageCollector(this);
```

Comment:
Because i wanted to have all job specific things encapsulated in the job, rather than make the GarbageCollector make too many assumptions about the way the job works and what kind of methods it uses to work.

Reply:
It's a method reference which is a more efficient way of writing a lambda (also much faster).

"SomeClass::methodName"

actually means 

"(myParam) -> SomeClass.methodName(myParam)"

And since it gets referenced inside the Job, the job has access to that private function.

What i want to achieve with this is: The GarbageCollector only needs to know which Jobs it supports - but it doesn't need to know anything about how the Job works and what requirements that job has - I still think this is a cleaner separation of logic than to hardcode all of the logic. into the GarbageCollector itself.

If we want to change the way the MilestonePrunerJob works, we only need to touch this file and the GarbageCollector can stay untouched an generic.